File: lua/neovim-pi/handshake.lua
-- Handshake: the first RPC exchange after pi attaches.
--
-- Pi calls `nvim_exec_lua` with this module's `exchange`
-- function, passing pi's version and capability list.
-- We record the pi channel and return our own.

local M = {}

-- Wire protocol version this plugin speaks.
local PROTOCOL_VERSION = "0.1.0"

-- Capabilities this plugin advertises.
local NVIM_CAPABILITIES = {
  "nvim.buffer.open",
  "nvim.buffer.markStale",
  "nvim.cursor.set",
  "nvim.cursor.get",
  "nvim.cursor.stream",
  "nvim.cursor.selection.get",
  "nvim.diff.files",
  "nvim.diff.pending",
  "nvim.diff.off",
  "nvim.extmark.set",
  "nvim.extmark.clear",
  "nvim.file.open",
  "nvim.file.save",
  "nvim.file.reload",
  "nvim.text.getRange",
  "nvim.text.setRange",
  "nvim.window.layout",
  "nvim.window.focus",
  "nvim.window.close",
  "nvim.buffer.list",
  "nvim.buffer.info",
  "nvim.buffer.switch",
  "nvim.buffer.delete",
  "nvim.status.publish",
  "nvim.lsp.query",
}

---@type { version: string, capabilities: string[] }?
local pi_peer = nil

--- Exchange capability info with pi. Called by pi via nvim_exec_lua.
---
--- Pi looks up its own channel id with `nvim_get_api_info`
--- before this call and passes it in explicitly so nvim can
--- call back into pi via `vim.rpcrequest` / `vim.rpcnotify`.
---@param pi_version string
---@param pi_caps string[]
---@param pi_channel integer
---@return { version: string, capabilities: string[] }
function M.exchange(pi_version, pi_caps, pi_channel)
  pi_peer = { version = pi_version, capabilities = pi_caps or {} }

  if pi_channel then
    require("neovim-pi.rpc").set_channel(pi_channel)
  end

  return {
    version = PROTOCOL_VERSION,
    capabilities = NVIM_CAPABILITIES,
  }
end

--- Inspect the recorded pi peer info.
function M.peer()
  return pi_peer and vim.deepcopy(pi_peer) or nil
end

--- Returns true when the pi peer advertised a capability.
---@param flag string
function M.pi_has(flag)
  if not pi_peer then
    return false
  end
  for _, c in ipairs(pi_peer.capabilities) do
    if c == flag then
      return true
    end
  end
  return false
end

return M

File: extensions/neovim-pi/src/handshake.ts
/**
 * Capability handshake.
 *
 * Right after attach, both sides advertise their feature
 * set. The handshake is a single RPC each way:
 *
 *   pi  → nvim : `pi.hello`    { version, capabilities }
 *   nvim → pi  : `nvim.hello`  { version, capabilities }
 *
 * Capabilities are additive flag names. Unknown flags
 * are silently ignored; missing flags degrade
 * gracefully. The spec for each flag lives in
 * `doc/capabilities.md`.
 */

import type { NeovimClient } from "neovim";

/** Wire version this implementation speaks. */
export const PROTOCOL_VERSION = "0.1.0";

/** Capabilities this pi extension advertises. */
export const PI_CAPABILITIES: readonly string[] = [
	"pi.session.get",
	"pi.session.subscribe",
	"pi.tool.list",
	"pi.tool.invoke",
	"pi.status.subscribe",
	"pi.prompt.append",
	"pi.buffer.uri.resolve",
];

/** Capabilities the nvim peer is expected to expose (best-effort check). */
export const EXPECTED_NVIM_CAPABILITIES: readonly string[] = [
	"nvim.buffer.open",
	"nvim.buffer.close",
	"nvim.buffer.markStale",
	"nvim.buffer.isModified",
	"nvim.buffer.reload",
	"nvim.window.cursor.set",
	"nvim.extmark.set",
	"nvim.extmark.clear",
	"nvim.file.open",
	"nvim.file.save",
	"nvim.text.getRange",
	"nvim.text.setRange",
	"nvim.window.layout",
	"nvim.buffer.list",
];

export interface PeerInfo {
	version: string;
	capabilities: string[];
}

let peer: PeerInfo | null = null;

/** Returns the peer's last-seen handshake info. */
export function getPeerInfo(): PeerInfo | null {
	return peer;
}

/**
 * Send our hello and receive nvim's. Throws on protocol mismatch.
 *
 * `nvim_get_api_info` returns `[channelId, apiMetadata]`. We pass
 * the channel id into the lua handshake so nvim can call back
 * into us via `vim.rpcrequest`/`rpcnotify`.
 */
export async function performHandshake(client: NeovimClient): Promise<PeerInfo> {
	const apiInfo = (await client.request("nvim_get_api_info", [])) as [number, unknown];
	const channelId = apiInfo[0];

	const response = (await client.request("nvim_exec_lua", [
		`return require("neovim-pi.handshake").exchange(...)`,
		[PROTOCOL_VERSION, PI_CAPABILITIES, channelId],
	])) as PeerInfo;

	if (!response?.version) {
		throw new Error("nvim handshake returned no version");
	}

	peer = response;
	return response;
}

/** Returns true when the peer advertised a capability. */
export function peerHas(capability: string): boolean {
	return peer?.capabilities.includes(capability) ?? false;
}
